client/network/src/service/out_events.rs

```diff
@@ -126,7 +123,14 @@ impl Stream for Receiver {
 
 	fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Event>> {
 		if let Some(ev) = ready!(Pin::new(&mut self.inner).poll_next(cx)) {
-			let _ = self.queue_size.fetch_sub(1, Ordering::Relaxed);
+			// `Ordering::Release` guarantees here that no loads (`poll_next()`) can be reordered
+			// after the `sub` operation. Assuming causality between `unbounded_send()` and
+			// `poll_next()` this ensures that counter decrement can't happen before the
+			// corresponding counter load in `send()`.
+			// Note that this is not the usual Release-Acquire ordering based on synchronization of
+			// store and load of an atomic variable, but more general application of Acquire and
+			// Release ordering properties.
+			let _ = self.queue_size.fetch_sub(1, Ordering::Release);
```

Comment:
I'd just use `SeqCst` instead to be on a safe size, I tried, but I don't think I understand all the implications of other options for something other than simple load or store operation.

Reply:
You usually can't go wrong with `SeqCst` for memory ordering. It's the default on most modern hardware, and pretty fast.

However, I don't understand the larger issue here. Reading the PR and the linked comments just made me confused. @dmitry-markin can you please summarize what the issue is we are trying to solve, and what memory ordering has to do with it?

Reply:
> However, I don't understand the larger issue here. Reading the PR and the linked comments just made me confused. @dmitry-markin can you please summarize what the issue is we are trying to solve, and what memory ordering has to do with it?

Originally we added the queue size monitoring to unbounded channels in order to detect message "leaks" — when some of the receivers are not polled and the messages keep stacking in the channel, "leaking" memory (we observed such bugs in the past). In the original implementation of this queue threshold warning I didn't bother at all about precise queue size estimation and memory order implications, considering that +/-1 error in queue size estimation is not a big deal when we talk about 100_000 or even 100 messages warning threshold.

It turned out, though, that at least some substrate users like @nazar-pc would like to set the exact warning threshold, specifically when dealing with channels with explicit acknowledgement (in this case the queue size threshold must be exactly 1), so I created this follow-up PR that makes the queue size measurement exact.

Nevertheless, doing things right with atomics seems not that trivial, especially when we go out of typical C++/Rust memory orderings territory and start to reason on a level of `loads` and `stores` and their reordering across the barrier (mentioned only in C++ docs). As the PR comment notes, we do not use atomics here to order other memory operations across the memory barrier imposed by the atomic operation, but actually do quite the opposite. Assuming the ordering of non-atomic (or other atomic) operations relative to our atomic operations, and the causality relation between pushing a message into the channel and popping it, we derive the ordering of our atomic operations relative to each other: the increment happens before the decrement, so we never hit the underflow of `usize` counter.

It's worth noting that `SeqCst` guarantee would only help as long as `mpsc` channel internally also uses `SeqCst`, because `SeqCst` ordering only guarantees single order of events relative to other `SeqCst` operations, and not to other atomic or non-atomic operations. So we can't really rely on this, because as though true right now, the use of `SeqCst` is the internal implementation detail of `mpsc` channel that can change in the future (especially taking into account that code comments there contradict the used `SeqCst` ordering).

Reply:
Honestly unless someone can point out some downsides I'd *probably* prefer if we'd just switch to a channel that supports checking the length natively instead of jerry rigging it ourselves with atomics and potentially depend on implicit implementation details of the current channel we're using.

e.g. the channel from `async-channel` [natively supports `.len()`](https://docs.rs/async-channel/1.8.0/async_channel/struct.Sender.html#method.len) (And I also like this crate for other reasons, e.g. it is a proper async channel which also supports sync usage, and can be switched to a bounded/unbounded variant and keep the same type. One downside it has is that it's a linked list internally so the performance can be bad, but we're using futures' channels now and it also uses a linked list, so there shouldn't really be a significant difference.)

Reply:
Just let's do what @koute is saying and continue living our happy lives without any synchronization problems :D